cmd/disk-cache.go

```diff
@@ -0,0 +1,674 @@
+	default:
+		err = e
+	}
+	if err != nil {
+		tmpPath := f.Name()
+		f.Close()
+		os.Remove(tmpPath)
+		return nil, err
+	}
+	return f, nil
+}
+
+// Returns the handle for the cached object
+func (c diskCache) Get(bucket, object string) (*os.File, ObjectInfo, bool, error) {
+	objectLock := globalNSMutex.NewNSLock(diskCacheLockingPrefix+bucket, object)
+	objectLock.RLock()
+	defer objectLock.RUnlock()
+
+	metaPath := c.encodedMetaPath(bucket, object)
+	objMeta := diskCacheObjectMeta{}
+	metaBytes, err := ioutil.ReadFile(metaPath)
+	if err != nil {
+		return nil, ObjectInfo{}, false, err
+	}
+	if err = json.Unmarshal(metaBytes, &objMeta); err != nil {
+		return nil, ObjectInfo{}, false, err
+	}
+	if objMeta.Version != diskCacheObjectMetaVersion {
+		return nil, ObjectInfo{}, false, errors.New("format not supported")
+	}
+	file, err := os.Open(c.encodedPath(bucket, object))
+	if err != nil {
+		return nil, ObjectInfo{}, false, err
+	}
+	c.updateAtime(bucket, object, time.Now().UTC())
+	return file, objMeta.toObjectInfo(), objMeta.Anonymous, nil
+}
+
+// Returns metadata of the cached object
+func (c diskCache) GetObjectInfo(bucket, object string) (ObjectInfo, bool, error) {
+	objectLock := globalNSMutex.NewNSLock(diskCacheLockingPrefix+bucket, object)
+	objectLock.RLock()
+	defer objectLock.RUnlock()
+
+	metaPath := c.encodedMetaPath(bucket, object)
+	objMeta := diskCacheObjectMeta{}
+	metaBytes, err := ioutil.ReadFile(metaPath)
+	if err != nil {
+		return ObjectInfo{}, false, err
+	}
+	if err := json.Unmarshal(metaBytes, &objMeta); err != nil {
+		return ObjectInfo{}, false, err
+	}
+
+	return objMeta.toObjectInfo(), objMeta.Anonymous, nil
+}
+
+// Deletes the cached object
+func (c diskCache) Delete(bucket, object string) error {
+	objectLock := globalNSMutex.NewNSLock(diskCacheLockingPrefix+bucket, object)
+	objectLock.Lock()
+	defer objectLock.Unlock()
+
+	if err := os.Remove(c.encodedPath(bucket, object)); err != nil {
+		return err
+	}
+	if err := os.Remove(c.encodedMetaPath(bucket, object)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Update atime for the cached object.
+func (c diskCache) updateAtime(bucket, object string, t time.Time) error {
+	entry := make(diskCacheBoltdbEntry)
+	entry.setAtime(t)
+	entryBytes, err := json.Marshal(entry)
+	if err != nil {
+		return err
+	}
+	return c.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(diskCacheBoltdbBucket))
+		return b.Put([]byte(pathJoin(bucket, object)), entryBytes)
+	})
+}
+
+// format.json structure
+// format.json is put in the root of the cache directory
+type diskCacheFormat struct {
+	Version string    `json:"version"`
+	Format  string    `json:"format"`
+	Time    time.Time `json:"createTime"`
+}
+
+// Inits the cache-dir if it is not init'ed already.
+// Initializing implies creation of cache-dir, data-dir, tmp-dir and format.json
+func newDiskCache(dir string, maxUsage, expiry int) (*diskCache, error) {
+	if err := os.MkdirAll(dir, 0766); err != nil {
```

Comment:
Is it allowed to have simultaneous minio process to write to same cache location? if not can we prevent it by holding lock properly like how we do in `fs` ?

Reply:
we do locking already in disk-cache

Reply:
can you point me to the right location?